### DISCUSSION TITLE
Modbus server - allowing only defined registers

### DISCUSSION THREAD

**Discussion by rwskinner** [pymodbus-dev/pymodbus#2268] [General]
if I only create 10 holding registers, my client can request say 50, and the first 10 register will have the proper values, and the rest are zeros.  The proper operation would be to return Illegal address  0x02 if any register outside the range of the 10 defined was requested.

They shouldn't have access to undefined registers.  
What is the best way to enforce this?  

Update:  I see it created 65535 regs by default.  So I guess h_regs_size  would do the trick.  Let me try.

**Comment by rwskinner** [comment]
if datablocks is created with TotalRegs=10, isn't that supposed to limit it to those 10 registers and return illegal address if any registers outside that range are requested?

    datablock = ModbusSequentialDataBlock(0x00, [0] * TotalRegs)
    context = ModbusSlaveContext(hr=datablock)
    store = ModbusSlaveContext(zero_mode=True)

    # Add datastores to slaves by slaveid
    slaves = {}
    slaves.update({SlaveId: store})
    context = ModbusServerContext(slaves=slaves, single=False)  #must have single at false in order to specifiy slaveid

**Comment by janiversen** [comment]
I am not sure that is the proper response, where in the standard did you read that!

IllegalAddress should only be returned if the parameter address is illegal not the count.

**Reply by rwskinner** [reply]
Its very well documented in the modbus protocol standard, and each device I've worked with in the industry adheres to it.
I'm actually an industrial communication engineer for 35 years, so yes, I've been around lots of modbus devices.

Notes in the example file says:
 # The datastores only respond to the addresses that are initialized
    # If you initialize a DataBlock to addresses of 0x00 to 0xFF, a request to
    # 0x100 will respond with an invalid address exception.
    # This is because many devices exhibit this kind of behavior (but not all)


With that said.  **My code was flawed** and when I fix it, it works properly.  I was overwriting my initial datablock I defined.

   datablock = ModbusSequentialDataBlock(0x00, [0] * TotalRegs)          
   store = ModbusSlaveContext(hr=datablock,zero_mode=True)         
    slaves = {}
    slaves.update({SlaveId: store})
    context = ModbusServerContext(slaves=slaves, single=False)  #must have single at false in order to specify slaveid
  
So now, if a client request a register out of the range, the server returns Illegal Address.  Perfect.

**Reply by janiversen** [reply]
Welcome to the club, I have worked with automation protocols for around 47 years and now I m retired.


I can tell you, that there also a lot of devices that handle this differently, a lot of chinese devices simply do mot return the full count. Fact is that the modbus standard do not define the handling properly.

**Reply by rwskinner** [reply]
Understood.  Most of my stuff has been name brand industrial controls.  I'm run across some "import" stuff that was crap as well but we try hard not to use them.  I have about 6000 edge device all over the US and very remote locations, so the hardware and software has to be very reliable.

The modbus protocol standards suck pretty bad. If you follow strict char timeouts then you'll typically end up with trouble, at least that's my experience.

Thanks for all the hard work and I appreciate the feedback.

**Reply by janiversen** [reply]
I was mostly in the other end, developing the firmware for the devices, mostly the protocol/message part but also SCADA systems. I used to work with bigger companies like siemens, so it's fair to say I have a different view than yours (different does not imply better or worse).

Thanks for using pymodbus

**Comment by janiversen** [comment]
You will find that real life devices respond very differently, but if you want the datastore to behave differently then pull requests are welcome.
